Run-Time Check Failure in system_rng on Windows (32 bit)

When compiling and running

```
$ python configure.py --debug-mode --cxxflags="/RTC1" --cpu=x86_32
$ jom -j 6
$ botan-test.exe system_rng
```

I get the error

> Run-Time Check Failure #0: The value of ESP was not properly saved across a function call.  This is usually a result of calling a function declared with one calling convention with a function pointer declared with a different calling convention.

![bildschirmfoto von 2018-05-07 18-22-19](https://user-images.githubusercontent.com/2603011/39712606-f1b0a8c8-5223-11e8-867d-217cc183da74.png)

The same happened when I ran my application in debugger, where I found that the line https://github.com/randombit/botan/blob/2.6.0/src/lib/rng/system_rng/system_rng.cpp#L44 triggers the run-time check failure.

The problem does not occur without `/RTC1` or in a 64 bit build.

[Looking at how Chromium uses RtlGenRandom](https://github.com/chromium/chromium/blob/9899ff1dc3/base/rand_util_win.cc), it seems possible and easier to use RtlGenRandom directly.

Adding NTAPI to the typedef is a syntax error:

```
src/lib/rng/system_rng/system_rng.cpp(53): error C2059: syntax error: '('
src/lib/rng/system_rng/system_rng.cpp(53): error C2238: unexpected token(s) preceding ';'
src/lib/rng/system_rng/system_rng.cpp(56): error C3646: 'm_rtlgenrandom': unknown override specifier
src/lib/rng/system_rng/system_rng.cpp(56): error C4430: missing type specifier - int assumed. Note: C++ does not support default-int
src/lib/rng/system_rng/system_rng.cpp(39): error C2065: 'm_rtlgenrandom': undeclared identifier
src/lib/rng/system_rng/system_rng.cpp(39): error C2065: 'RtlGenRandom_f': undeclared identifier
src/lib/rng/system_rng/system_rng.cpp(39): error C2672: 'Botan::Dynamically_Loaded_Library::resolve': no matching overloaded function found
src/lib/rng/system_rng/system_rng.cpp(39): error C2974: 'Botan::Dynamically_Loaded_Library::resolve': invalid template argument for 'T', type expected
build\include\botan/dyn_load.h(53): note: see declaration of 'Botan::Dynamically_Loaded_Library::resolve'
src/lib/rng/system_rng/system_rng.cpp(44): error C3861: 'm_rtlgenrandom': identifier not found
jom: C:\Users\Kullo\botan\Makefile [build\obj\lib\rng_system_rng.obj] Error 2
        cl /MDd /bigobj /DBOTAN_DLL=__declspec(dllexport) /EHs /GR /RTC1 /W4 /wd4250 /wd4251 /wd4275  /Ibuild\include /Ibuild\include\external /nologo /c src/lib/stream/rc4/rc4.cpp /Fobuild\obj\lib\stream_rc4.obj
rc4.cpp
```

I tries to use BOOLEAN instead of BOOL for the return type (as this is how it is documented, but nothing changes.

Then I some sizeof tests for the used types on 64 and 32 bit:

```
64bit> sizes.exe
ULONG 4
std::size_t 8

PVOID 8
uint8_t* 8

BOOL 4
BOOLEAN 1
TRUE 4
FALSE 4
```

```
32bit> sizes.exe
ULONG 4
std::size_t 4

PVOID 4
uint8_t* 4

BOOL 4
BOOLEAN 1
TRUE 4
FALSE 4
```

For some reason the mismatch between ULONG and std::size_t on 64 bit is not a problem.

> I didn't know about /RTC flag. We should probably enable that with the debug/sanitizer build.

Neither did I. cmake added it for me in the debug configuration. The problem with `/RTC1` is that it only workes with no optimization (`/Od`) which was considered too slow for CI testing a while ago.

--- a/src/lib/rng/system_rng/system_rng.cpp
+++ b/src/lib/rng/system_rng/system_rng.cpp
@@ -50,7 +50,7 @@ class System_RNG_Impl final : public RandomNumberGenerator
       void clear() override { /* not possible */ }
       std::string name() const override { return "RtlGenRandom"; }
    private:
-      typedef BOOL (*RtlGenRandom_f)(PVOID, ULONG);
+      typedef BOOL (NTAPI *RtlGenRandom_f)(PVOID, ULONG);
 
       Dynamically_Loaded_Library m_advapi;
       RtlGenRandom_f m_rtlgenrandom;

--- a/src/lib/rng/system_rng/system_rng.cpp
+++ b/src/lib/rng/system_rng/system_rng.cpp
@@ -41,7 +41,8 @@ class System_RNG_Impl final : public RandomNumberGenerator
 
       void randomize(uint8_t buf[], size_t len) override
          {
-         if(m_rtlgenrandom(buf, len) == false)
+         bool success = m_rtlgenrandom(buf, len) == TRUE;
+         if(!success)
             throw Exception("RtlGenRandom failed");
          }
 

--- a/src/lib/rng/system_rng/system_rng.cpp
+++ b/src/lib/rng/system_rng/system_rng.cpp
@@ -36,7 +36,7 @@ class System_RNG_Impl final : public RandomNumberGenerator
       System_RNG_Impl() : m_advapi("advapi32.dll")
          {
          // This throws if the function is not found
-         m_rtlgenrandom = m_advapi.resolve<RtlGenRandom_f>("SystemFunction036");
+         m_rtlgenrandom = m_advapi.resolve<RtlGenRandom_fptr>("SystemFunction036");
          }
 
       void randomize(uint8_t buf[], size_t len) override
@@ -51,10 +51,10 @@ class System_RNG_Impl final : public RandomNumberGenerator
       void clear() override { /* not possible */ }
       std::string name() const override { return "RtlGenRandom"; }
    private:
-      using RtlGenRandom_f = BOOLEAN (NTAPI *)(PVOID, ULONG);
+      using RtlGenRandom_fptr = BOOLEAN (NTAPI *)(PVOID, ULONG);
 
       Dynamically_Loaded_Library m_advapi;
-      RtlGenRandom_f m_rtlgenrandom;
+      RtlGenRandom_fptr m_rtlgenrandom;
    };
 
 #elif defined(BOTAN_TARGET_OS_HAS_ARC4RANDOM)

--- a/src/lib/rng/system_rng/system_rng.cpp
+++ b/src/lib/rng/system_rng/system_rng.cpp
@@ -51,7 +51,7 @@ class System_RNG_Impl final : public RandomNumberGenerator
       void clear() override { /* not possible */ }
       std::string name() const override { return "RtlGenRandom"; }
    private:
-      typedef BOOLEAN (NTAPI *RtlGenRandom_f)(PVOID, ULONG);
+      using RtlGenRandom_f = BOOLEAN (NTAPI *)(PVOID, ULONG);
 
       Dynamically_Loaded_Library m_advapi;
       RtlGenRandom_f m_rtlgenrandom;

--- a/src/lib/rng/system_rng/system_rng.cpp
+++ b/src/lib/rng/system_rng/system_rng.cpp
@@ -50,7 +50,7 @@ class System_RNG_Impl final : public RandomNumberGenerator
       void clear() override { /* not possible */ }
       std::string name() const override { return "RtlGenRandom"; }
    private:
-      typedef BOOL (NTAPI *RtlGenRandom_f)(PVOID, ULONG);
+      typedef BOOLEAN (NTAPI *RtlGenRandom_f)(PVOID, ULONG);
 
       Dynamically_Loaded_Library m_advapi;
       RtlGenRandom_f m_rtlgenrandom;

--- a/src/lib/rng/system_rng/system_rng.cpp
+++ b/src/lib/rng/system_rng/system_rng.cpp
@@ -51,7 +51,10 @@ class System_RNG_Impl final : public RandomNumberGenerator
       void clear() override { /* not possible */ }
       std::string name() const override { return "RtlGenRandom"; }
    private:
-      using RtlGenRandom_fptr = BOOLEAN (NTAPI *)(PVOID, ULONG);
+      // Use type BYTE instead of BOOLEAN because of a naming conflict
+      // https://msdn.microsoft.com/en-us/library/windows/desktop/aa387694(v=vs.85).aspx
+      // https://msdn.microsoft.com/en-us/library/windows/desktop/aa383751(v=vs.85).aspx
+      using RtlGenRandom_fptr = BYTE (NTAPI *)(PVOID, ULONG);
 
       Dynamically_Loaded_Library m_advapi;
       RtlGenRandom_fptr m_rtlgenrandom;

